ops/services/alerts/app_service_metrics/main.tf

```diff
@@ -273,3 +273,62 @@ ${local.skip_on_weekends}
     action_group = var.action_group_ids
   }
 }
+
+resource "azurerm_monitor_scheduled_query_rules_alert" "batched_uploader_single_failure_detected" {
+  name                = "${var.env}-batched-uploader-single-failure-detected"
+  description         = "QueueBatchedReportStreamUploader failed to successfully complete"
+  location            = data.azurerm_resource_group.app.location
+  resource_group_name = var.rg_name
+  severity            = var.severity
+  frequency           = 5
+  time_window         = 5
+  enabled             = contains(var.disabled_alerts, "batched_uploader_single_failure_detected") ? false : true
+
+  data_source_id = var.app_insights_id
+
+  query = <<-QUERY
+dependencies
+${local.skip_on_weekends}
+| where timestamp >= ago(7m) 
+    and operation_Name =~ 'QueueBatchedReportStreamUploader' 
+    and success != true
+  QUERY
+
+  trigger {
+    operator  = "GreaterThan"
+    threshold = 0
+  }
+
+  action {
+    action_group = var.action_group_ids
+  }
+}
+
+resource "azurerm_monitor_scheduled_query_rules_alert" "batched_uploader_function_not_triggering" {
+  name                = "${var.env}-batched-uploader-function-not-triggering"
```

Comment:
I have a little bit of hesitancy around this alert - what happens if there's nothing in the queue to be processed? Will the function still trigger, see there's nothing to offload, and pass a success flag? Or will it not trigger at all?

This likely wasn't an issue before when we were sending results every 15 minutes, but now that we're down to 2 mins I could see some middle-of-the-night empty queues...which is exactly when I wouldn't want this to fire 😄

Reply:
Ooh. That is...ooh. Great thought.

So, from what I can see, we trigger this off of a timer. Meaning, it is going to trigger regardless. From what I can see, we attempt a dequeue here: https://github.com/CDCgov/prime-simplereport/blob/772ae2bf8abc8ae0480276d50316942c2f1c81cc/ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts#L35

It's my interpretation that, if there are no messages to dequeue, the flow will proceed as expected, but an empty csv will be created. We should still see successful invocations in Application Insights.

The _only_ time this should theoretically fire is if the uploader is completely missing.

Please let me know if you see a flaw with what I've posted, though! I definitely don't want false alerts getting fired off. 😅

Reply:
I think we might be able to use metrics from the function on this one - it tracks the # of executions over a given window.

Reply:
I previously commented that we could use the # of function executions as a metric but I had to delete that right away, because it turns out that's not true - that metric is real, but it's not triggered on every execution for some reason, so I'm unclear what it's actually tracking.

But! I can confirm that the following query works, even run for a span between 3 and 4AM ET:

```
traces
| where timestamp between(datetime("2022-02-09 08:00") .. datetime("2022-02-09 09:00"))
    and operation_Name =~ 'QueueBatchedReportStreamUploader'
```

Note the change to `traces` - using `dependencies` isn't generating any results for me for some reason.

Reply:
Oh yeah that's a good point. I agree, I think `requests` makes sense because these alerts combined will catch the two cases we're looking for.

Reply:
This looks good to me, thank you both for digging into it! I also agree that using `requests` seems like the right move here.